<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv("sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
9    0.0164  0.0173  0.0347  0.0070  0.0187  0.0671  0.1056  0.0697  0.0962   
197  0.0366  0.0421  0.0504  0.0250  0.0596  0.0252  0.0958  0.0991  0.1419   
120  0.0346  0.0509  0.0079  0.0243  0.0432  0.0735  0.0938  0.1134  0.1228   
50   0.0353  0.0713  0.0326  0.0272  0.0370  0.0792  0.1083  0.0687  0.0298   
125  0.0228  0.0853  0.1000  0.0428  0.1117  0.1651  0.1597  0.2116  0.3295   

         9   ...      51      52      53      54      55      56      57  \
9    0.0251  ...  0.0090  0.0223  0.0179  0.0084  0.0068  0.0032  0.0035   
197  0.1847  ...  0.0132  0.0027  0.0022  0.0059  0.0016  0.0025  0.0017   
120  0.1508  ...  0.0040  0.0122  0.0107  0.0112  0.0102  0.0052  0.0024   
50   0.0880  ...  0.0163  0.0242  0.0043  0.0202  0.0108  0.0037  0.0096   
125  0.3517  ...  0.0172  0.0191  0.0260  0.0140  0.0125  0.0116  0.0093   

         58      59  60  
9    0.0056  0.0040   R  
197  0.0027  0.0027   M  
120  0.0079  0.0031   M  
50   0.0093  0.0053   R  
125  0.0012  0.0036   M  

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

In [5]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [7]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
y = pd.get_dummies(y, drop_first=True)#we drop 1st column which is M because only r is sufficient 
y.sample(5) # R --> 1 and M --> 0

R
203  0
41   1
103  0
67   1
108  0

In [10]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [11]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [13]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

# Using Deep Learning Model

## Model without Dropout Layer

In [14]:
import tensorflow as tf
from tensorflow import keras

In [15]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)#mini gradient descent

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.7044 - accuracy: 0.4872
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6728 - accuracy: 0.5449
Epoch 3/100
20/20 [==============================] - 0s 1ms/step - loss: 0.6517 - accuracy: 0.6090
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6312 - accuracy: 0.7692
Epoch 5/100
20/20 [==============================] - 0s 1ms/step - loss: 0.5968 - accuracy: 0.6923
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5734 - accuracy: 0.7949
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5337 - accuracy: 0.7821
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4952 - accuracy: 0.8141
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4693 - accuracy: 0.8141
Epoch 10/100
20/20 [==============================] - 0s 3ms/step - loss: 0.4369 - accuracy: 0.8526
Epoch 11/

20/20 [==============================] - 0s 3ms/step - loss: 0.0033 - accuracy: 1.0000
Epoch 83/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0030 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0032 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0028 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0025 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0025 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0024 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0023 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0023 - accuracy: 1.0000
Epoch 91/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0022 - accuracy: 1.0000
Epoch 92/100


In [16]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 16ms/step - loss: 1.1135 - accuracy: 0.7308


[1.1135145425796509, 0.7307692170143127]

Training Accuracy >>> Test Accuracy
Overfitting occurs

In [17]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[4.4015056e-08 9.8791832e-01 9.8940599e-01 4.7323874e-06 9.9999797e-01
 9.9997640e-01 8.6086249e-01 1.0000000e+00 3.5491150e-05 9.9999976e-01]
[0. 1. 1. 0. 1. 1. 1. 1. 0. 1.]


In [18]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [19]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.81      0.76        27
           1       0.76      0.64      0.70        25

    accuracy                           0.73        52
   macro avg       0.74      0.73      0.73        52
weighted avg       0.73      0.73      0.73        52



### Model with Dropout Layer

In [20]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 3ms/step - loss: 0.7201 - accuracy: 0.5256
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7021 - accuracy: 0.5256
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7213 - accuracy: 0.5192
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7046 - accuracy: 0.5128
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7191 - accuracy: 0.5192
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6966 - accuracy: 0.5064
Epoch 7/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7116 - accuracy: 0.4487
Epoch 8/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6950 - accuracy: 0.4808
Epoch 9/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6962 - accuracy: 0.4808
Epoch 10/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6784 - accuracy: 0.5897
Epoch 11/

20/20 [==============================] - 0s 3ms/step - loss: 0.3361 - accuracy: 0.8846
Epoch 84/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3801 - accuracy: 0.8526
Epoch 85/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3810 - accuracy: 0.8590
Epoch 86/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3337 - accuracy: 0.8782
Epoch 87/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3458 - accuracy: 0.8782
Epoch 88/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3524 - accuracy: 0.9103
Epoch 89/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3421 - accuracy: 0.8590
Epoch 90/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3330 - accuracy: 0.8974
Epoch 91/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3080 - accuracy: 0.8846
Epoch 92/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3142 - accuracy: 0.8654
Epoch 93/100


In [21]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 18ms/step - loss: 0.5188 - accuracy: 0.7692


[0.5188463926315308, 0.7692307829856873]

Training Accuracy is still good but Test Accuracy Improved

In [ ]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

In [ ]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**